src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    const addressTypeErrors  = this.validateAddressTypes(addresses.primary, addresses.secondary, writeToUsTypeId);
+    const addressesValid = primaryValidationResult.addressValid && primaryValidationResult.postcodeValid &&
+      secondaryValidationResult.addressValid && secondaryValidationResult.postcodeValid && addressTypeErrors.length === 0;
+
+    if (!addressesValid) {
+      const allErrors = primaryValidationResult.errors.concat(secondaryValidationResult.errors).concat(addressTypeErrors);
+      await this.render(req, res, false, addresses, allErrors,
+        !primaryValidationResult.postcodeValid, !secondaryValidationResult.postcodeValid);
+      return;
+    }
+
+    // Post addresses to API if valid
+    await req.scope.cradle.api.updateCourtAddresses(req.params.slug, this.convertToApiType(addresses))
+      .then(async (addressList: CourtAddress[]) => await this.render(req, res, true, this.convertToDisplayAddresses(addressList)) )
+      .catch(async (reason: AxiosError) => {
+        if (reason.response.status === 400) {
+          const postcodeValidation = this.checkErrorResponseForPostcodeErrors(reason, addresses);
+          const errors = postcodeValidation.errors.length === 0
+            ? [this.updateAddressError] // we've encountered a 400 for a reason other than postcodes
+            : postcodeValidation.errors;
+          await this.render(req, res, false, addresses, errors, postcodeValidation.primaryInvalid, postcodeValidation.secondaryInvalid);
+        } else {
+          await this.render(req, res, false, addresses, [this.updateAddressError]);
+        }
+      });
+  }
+
+  private async render(
+    req: AuthedRequest,
+    res: Response,
+    updated = false,
+    addresses: DisplayCourtAddresses = null,
+    errorMsgs: string[] = [],
+    primaryPostcodeInvalid = false,
+    secondaryPostcodeInvalid = false) {
+
+    const slug: string = req.params.slug as string;
+    let fatalError = false;
+
+    if (!addresses) {
+      await req.scope.cradle.api.getCourtAddresses(slug)
+        .then((addressList: CourtAddress[]) => addresses = this.convertToDisplayAddresses(addressList))
+        .catch(() => {
+          errorMsgs.push(this.getAddressesError);
+          fatalError = true;
+        });
+    }
+
+    let addressTypes: AddressType[] = [];
+    await req.scope.cradle.api.getAddressTypes()
+      .then((types: AddressType[]) => addressTypes = types)
+      .catch(() => {
+        errorMsgs.push(this.getAddressTypesError);
+        fatalError = true;
+      });
+    const writeToUsTypes = addressTypes.filter(at => at.name.toLowerCase() === this.writeToUsAddressType.toLowerCase()).map(at => at.id);
+
+    const pageData: CourtAddressPageData = {
+      addressTypesPrimary: this.getAddressTypesForSelect(addressTypes, true),
+      addressTypesSecondary: this.getAddressTypesForSelect(addressTypes, false),
+      addresses: addresses,
+      writeToUsTypeId: writeToUsTypes.length === 1 ? writeToUsTypes[0] : null,
```

Comment:
Is there a way for us to not use null here if we can help it? Not too sure if it matters as much for the frontend as it does for the backend; one to consider

Reply:
I think null is fine here. Otherwise, we'd need to give it some default value.